stablehlo/dialect/TypeInference.h

```diff
@@ -83,19 +83,60 @@ LogicalResult verifyReducerShape(Optional<Location> loc, Block& block,
                                  ArrayRef<int64_t> allowedDimensions,
                                  bool allInputsUnranked);
 
-// Verifies replica groups attached to collective communication operations.
-// P1. 'replicaGroups' must be a 2-D tensor.
-// P2. replicaGroups' cannot be empty.
-// P3. If `allGroupsMustHaveSameSize` is true, then each group is of the same
-//     size.
-// P4. All values in `replica_groups` are unique and covers all the values in
-//     the interval [0, N-1], where N is the total number of replica ids.
-// P5. replica group size must be equal to 'expectedGroupSize'.
-LogicalResult verifyReplicaGroups(Optional<Location> location,
-                                  DenseIntElementsAttr replicaGroups,
-                                  bool allGroupsMustHaveSameSize,
-                                  bool useGlobalDeviceIds,
-                                  Optional<size_t> expectedGroupSize);
+template <typename dimTy>
+static void inferGatherShape(
```

Comment:
Given that the only template instantiation used outside TypeInference.cpp is `dimTy = Value`, let's have a regular function `inferGatherShape` with `dimTy = Value` in the header file and then sort out the rest inside TypeInference.cpp.

Reply:
Thanks for this suggestions! While I'm not sure about this: this a large function. If we decide to keep the `reifyReturnTypeShapes()` in https://github.com/openxla/stablehlo/issues/828, then it's better to keep this shared template function here instead of duplicating the same long code both in `TypeInference.h` (used by reify) and in `TypeInference.cpp`.
If `reifyReturnTypeShapes()` will be gone, then this function will definitely be deleted and moved to inline in `TypeInference.cpp`.
I prefer do this after https://github.com/openxla/stablehlo/issues/828. What do you think of this?

Reply:
More concretely, here's what I'm proposing we do in this PR:

```C++
// In TypeInference.h
void reifyGatherDimSizes(
    int64_t resultRank, llvm::function_ref<Value(int64_t)> getStartIndicesDim,
    llvm::function_ref<Value(int64_t)> getSliceDim,
    ArrayRef<int64_t> offsetDims, ArrayRef<int64_t> collapsedSliceDims,
    ArrayRef<int64_t> startIndexMap, int64_t indexVectorDim,
    SmallVectorImpl<Value>& shape);

// In TypeInference.cpp
template <typename dimTy>
void inferGatherShape(...)

void reifyGatherDimSizes(
    int64_t resultRank, llvm::function_ref<Value(int64_t)> getStartIndicesDim,
    llvm::function_ref<Value(int64_t)> getSliceDim,
    ArrayRef<int64_t> offsetDims, ArrayRef<int64_t> collapsedSliceDims,
    ArrayRef<int64_t> startIndexMap, int64_t indexVectorDim,
    SmallVectorImpl<Value>& shape) {
  inferGatherShape(resultRank, ...);
}
```

This will keep TypeInference.h nice & clean and will keep full flexibility for whatever we decide in #828.

Reply:
Oh I see. This indeed make header clean, thanks! Please see this in the last commit.